docprompt/storage.py

```diff
@@ -0,0 +1,242 @@
+
+    @computed_field
+    @property
+    def pdf(self) -> str:
+        """The path for the PDF file."""
+        return f"{self.base_path}/{self.file_hash}/base.pdf"
+
+    @computed_field
+    @property
+    def metadata(self) -> str:
+        """The path for the metadata file."""
+        return f"{self.base_path}/{self.file_hash}/base.json"
+
+    @computed_field
+    @property
+    def page_metadata(self) -> str:
+        """The path for the page metadata file."""
+        return f"{self.base_path}/{self.file_hash}/pages.json"
+
+
+class FileSystemAnnotation:
+    """FileSystemType is a custom type for fsspec.AbstractFileSystem instances.
+
+    This class handles defining the necessary information for pydantic to be able to validate
+    and fsspec file system as a field in a Pydanitc model.
+    """
+
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
+        """Validate that it is a valid fsspec file system."""
+        if not isinstance(v, fsspec.AbstractFileSystem):
+            raise TypeError("Must be an fsspec.AbstractFileSystem instance")
+        return v
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls, _source_type: Any, _handler: Any
+    ) -> core_schema.CoreSchema:
+        """Get the Pydantic Core Schema for the FileSystemType."""
+        return core_schema.no_info_after_validator_function(
+            cls.validate,
+            core_schema.any_schema(),
+            serialization=core_schema.to_string_ser_schema(),
+        )
+
+
+class FileSystemManager(BaseModel):
+    """The FileSystemManager provides a wrapper around fsspec to provide a clean interface for
+    reading and writing sidecar directories for storing document nodes.
+    """
+
+    path: str = Field(...)
+    fs: FileSystemAnnotation = Field(...)
+    fs_kwargs: Dict[str, Any] = Field(...)
+
+    def __init__(self, url: str, **kwargs: Any) -> None:
+        """Initialize the FileSystemManager with the path and file system backend."""
+        super().__init__(path=url, fs_kwargs=kwargs)
+
+    @model_validator(mode="before")
+    @classmethod
+    def validate_filesystem_protocol_and_kwargs(cls, data: Any) -> Any:
+        """Validate that the path is a valid filesystem path."""
+
+        path = data.get("path", None)
+        fs_kwargs = data.get("fs_kwargs", {})
+
+        # Validate that the path is a valid filesystem path
+        file_system = fsspec.url_to_fs(path, **fs_kwargs)
+
+        # Set the data values on the model
+        data["fs"] = file_system[0]
+        data["path"] = file_system[1]
+        data["fs_kwargs"] = fs_kwargs
+
+        return data
+
+    @model_validator(mode="after")
+    def validate_base_path_exists(self) -> Self:
+        """Validate that the base path exists after instantiating the file system."""
+        if not self.fs.exists(self.path, **self.fs_kwargs):
+            self.fs.mkdirs(self.path, **self.fs_kwargs)
```

Comment:
This means a filesystem is modified as a side effect of instantiation. Should this only be an explicit effect of write operations?

Reply:
Yes, with new implementation of private `_wrtie` method, we can probably drop this post validator. Will remove and update unit tests as well to reflect changes